### ISSUE TITLE
Log Loss Custom Objective Python Correct?

### ISSUE THREAD

**Issue by BrianMiner** [dmlc/xgboost#1060]
Sorry if this is a silly question...but in the example here https://github.com/dmlc/xgboost/blob/master/demo/guide-python/custom_objective.py

shouldn't the gradient be (label-preds) instead of (preds - labels) for log loss?

**Comment by tqchen** [comment]
It should be preds - label. As logloss is define as negation of log likelihood. The algorithm will do gradient descent instead of ascent

**Comment by BrianMiner** [comment]
This had stuck in my mind: https://gradientboostedmodels.googlecode.com/git/gbm/inst/doc/gbm.pdf from the R GBM package. Where on page 10 they give the gradient as (label - preds). But looking back I see the deviance differs in sign from the log loss. Any idea about this? A different formulation?

**Comment by tqchen** [comment]
I guess it is only different formalization. In R's GBM uses maximum log-likelihood as objective, and take on ascenting direction of gradient, while xgboost uses minimize negation of likelihood, and take descenting direction of gradient.

Both are consistent with each other. Though the definition of objective differs by a negation sign

**Comment by BrianMiner** [comment]
I think this explains the formulation difference for future readers: http://stats.stackexchange.com/questions/157870/scikit-binomial-deviance-loss-function

In the R version, they label z_i as the gradient on page 10 but in the actually algorithm (Friedmans) on page 2, it is rightfully called the negative gradient. So it would seem they are moving in the opposite direction of the gradient as well.
